File: src/main/java/com/refinedmods/refinedstorage/energy/ItemEnergyStorage.java
package com.refinedmods.refinedstorage.energy;

import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.energy.EnergyStorage;

public class ItemEnergyStorage extends EnergyStorage {
    private static final String NBT_ENERGY = "Energy";

    private final ItemStack stack;

    public ItemEnergyStorage(ItemStack stack, int capacity) {
        super(capacity, Integer.MAX_VALUE, Integer.MAX_VALUE);

        this.stack = stack;
        this.energy = stack.hasTag() && stack.getTag().contains(NBT_ENERGY) ? stack.getTag().getInt(NBT_ENERGY) : 0;
    }

    @Override
    public int receiveEnergy(int maxReceive, boolean simulate) {
        int received = super.receiveEnergy(maxReceive, simulate);

        if (received > 0 && !simulate) {
            if (!stack.hasTag()) {
                stack.setTag(new CompoundTag());
            }

            stack.getTag().putInt(NBT_ENERGY, getEnergyStored());
        }

        return received;
    }

    @Override
    public int extractEnergy(int maxExtract, boolean simulate) {
        int extracted = super.extractEnergy(maxExtract, simulate);

        if (extracted > 0 && !simulate) {
            if (!stack.hasTag()) {
                stack.setTag(new CompoundTag());
            }

            stack.getTag().putInt(NBT_ENERGY, getEnergyStored());
        }

        return extracted;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/energy/ItemEnergyStorage.kt
package com.refinedmods.refinedstorage.energy

import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraftforge.energy.EnergyStorage

class ItemEnergyStorage(private val stack: ItemStack, capacity: Int) : EnergyStorage(capacity, Int.MAX_VALUE, Int.MAX_VALUE) {
    fun receiveEnergy(maxReceive: Int, simulate: Boolean): Int {
        val received: Int = super.receiveEnergy(maxReceive, simulate)
        if (received > 0 && !simulate) {
            if (!stack.hasTag()) {
                stack.tag = CompoundTag()
            }
            stack.tag!!.putInt(NBT_ENERGY, getEnergyStored())
        }
        return received
    }

    fun extractEnergy(maxExtract: Int, simulate: Boolean): Int {
        val extracted: Int = super.extractEnergy(maxExtract, simulate)
        if (extracted > 0 && !simulate) {
            if (!stack.hasTag()) {
                stack.tag = CompoundTag()
            }
            stack.tag!!.putInt(NBT_ENERGY, getEnergyStored())
        }
        return extracted
    }

    companion object {
        private const val NBT_ENERGY = "Energy"
    }

    init {
        this.energy = if (stack.hasTag() && stack.tag!!.contains(NBT_ENERGY)) stack.tag!!.getInt(NBT_ENERGY) else 0
    }
}